src/merge_zip_files.sh

```diff
@@ -0,0 +1,76 @@
+#!/bin/bash
+
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# A script that zips the content of the inputs zip files under a given directory
+# structure in the output zip file. "-" can be passed if no top-level directory
+# structure is required.
+
+# Usage: third_party/merge_zip_files.sh directory_prefix output_zip [input_zip_files]
+#
+# For example, if we have the following zips and their content:
+# a.zip:
+#    dir1/a1.cc
+#    a2.cc
+# b.zip:
+#    dir2/b1.cc
+#    b2.cc
+#
+# third_party_zip_files.sh src/main/cpp my_archive.zip a.zip b.zip
+# will create the archive my_archive.zip containing:
+# src/main/cpp/a2.cc
+# src/main/cpp/b2.cc
+# src/main/cpp/dir1/a1.cc
+# src/main/cpp/dir2/b1.cc
+#
+# third_party_zip_files.sh - my_archive.zip a.zip b.zip
+# will create the archive my_archive.zip containing:
+# a2.cc
+# b2.cc
+# dir1/a1.cc
+# dir2/b1.cc
+
+set -euo pipefail
+
+directory_prefix="$1"; shift
+output="$1"; shift
+
+initial_pwd="$(pwd)"
+
+tmp_dir=$(mktemp -d -t 'tmp_bazel_zip_files_XXXXX')
+trap "rm -fr $tmp_dir" EXIT
+tmp_zip="$tmp_dir/archive.zip"
+
+if [[ "$directory_prefix" == "-" ]]; then
+  for curr_zip in "$@"
```

Comment:
nit: why the if? you could accomplish this simply by

```
if [[ "$directory_prefix" == "-" ]]; then
  directory_prefix=""
fi

cd "$tmp_dir"
mkdir -p "$tmp_dir/$directory_prefix"
for curr_zip in "$@"
do
   unzip -q "$curr_zip" -d "$tmp_dir/$directory_prefix"
done
```

or, if the trailing slash makes a difference, by adding that to `$directory_prefix`, i.e.

```
if [[ <no directory prefix >>; then
  directory_suffix=""
else
  directory_suffix="/$directory_prefix"
fi
```

Reply:
The trailing slash is not the only thing to consider. Unzipping is ok, but zipping back is the problem. We either have to zip directory_prefix to preserve the directory_prefix in the zip, or all the content in the current directory (`.`). Hacking all these differences leads to unreadable bash code, making it easy to forget what slash and `.` comes where.